### PR TITLE
Always set RevisionGrid RowCount on LoadingCompleted

### PR DESCRIPTION
**Pull request by mstv**
Fixes https://github.com/gitextensions/gitextensions/issues/11065#issuecomment-1599591171
Fixes https://github.com/gitextensions/gitextensions/pull/11046#issuecomment-1595355058

## Proposed changes

- Always set `RevisionGrid.RowCount` to `graph.Count` on `LoadingCompleted`

## Test methodology 

- manual (reproducible with a private repo)

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs [hunk 1]
```diff
@@ -425,13 +425,19 @@ public void LoadingCompleted()
...
+                MarkAsDataLoadingComplete();
+            }
+            else
             {
                 // Rows have not been selected yet
                 ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                 {
                     await this.SwitchToMainThreadAsync();
 
+                    SetRowCountAndSelectRowsIfReady();
```
Comment (gerhardol) [inline GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:439]: This line vs old 466 is the actual change in the PR.
Is it OK to update RowCount before rendering is initiated?

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs [hunk 2]
```diff
@@ -425,13 +425,19 @@ public void LoadingCompleted()
                 _loadedToBeSelectedRevisionsCount = ToBeSelectedObjectIds.Count;
             }
 
-            if (_loadedToBeSelectedRevisionsCount > 0 && _revisionGraph.Count > 0)
+            if (_revisionGraph.Count == 0)
```
Comment (gerhardol) [inline GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:428]: If there is nothing to select, the loop is not needed.
Or did you have any other reason?

```suggestion
            if (_revisionGraph.Count == 0 || _loadedToBeSelectedRevisionsCount == 0)
```

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs [hunk 3]
```diff
@@ -476,10 +477,6 @@ public void LoadingCompleted()
                 })
                 .FileAndForget();
             }
-            else
-            {
-                MarkAsDataLoadingComplete();
-            }
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:482]: Could you please explain why this is being removed?

Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:482]: It is moved to line 430

Comment (RussKie) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:482]: Thank you, somehow I missed that.

**Review by RussKie — COMMENTED** [review]
I don't really understand this. If it's working - :shipit:

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs (hunk 2 above)
Comment (mstv) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:428]: The issue is caused by not setting `RowCount` in some cases. That's why this condition should be independent of all these selection things.
`RowCount` can only be set after switching to the main thread.

> If there is nothing to select, the loop is not needed.

Me too thought that the while loop below might have become superfluous after calling `SetRowCountAndSelectRowsIfReady()`. Though I was not sure and did not want to dig deeper.

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs (hunk 1 above)
Comment (mstv) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:439]: > Is it OK to update RowCount before rendering is initiated?

Yes, it is. `RowCount` is unconditionally set to the number of loaded revisions.
When rendering is requested, `CacheTo()` is called. This function creates and sorts the graph segments as far a needed.

File: GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs (hunk 2 above)
Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:428]: > RowCount can only be set after switching to the main thread.

Then set RowCount early in the async code and exit before the while if no revisions at all or none to select?

Comment (mstv) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:428]: I'd say that's what is done. What have I missed?
```
            if (_revisionGraph.Count == 0)
            {
                MarkAsDataLoadingComplete();
            }
            else
            {
                // Rows have not been selected yet
                ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                {
                    await this.SwitchToMainThreadAsync();

                    SetRowCountAndSelectRowsIfReady();

                    if (_toBeSelectedGraphIndexesCache.Value.Count == 0)
                    {
                        // Nothing to select or interrupted
                        MarkAsDataLoadingComplete();
                        return;
                    }
```

Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs:428]: Yes, it seem to work
The loading is complex, I need to look really close at it even if I have been involved in this part.

**Review by gerhardol — APPROVED** [review]

**Pull request merged**